### PR TITLE
[TASK] Allow functions in samplers/optimizers to be implemented in the same python module

### LINKED ISSUE #2300: [TASK] Allow implementation of functions in the same module for samplers/optimizers
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**

For cases where a lot of variables need to be generated using functions, the current approach is to create a single function python module for each variable. For a lot of variables, this will require the creation of tens of files that are difficult to track and maintain (overall in automated workflows)

**Describe the solution you'd like**

It would be convenient to allow for the  implementation of functions in the same module for samplers/optimizers.

**Describe alternatives you've considered**

Usage of current approach.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [ ] 1. Is it tagged with a type: defect or task?
- [ ] 2. Is it tagged with a priority: critical, normal or minor?
- [ ] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [ ] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [ ] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [ ] 1. If the issue is a defect, is the defect fixed?
- [ ] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [ ] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [ ] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [ ] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by alfoa**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes  #2300

##### What are the significant changes in functionality due to this change request?
This PR Allows functions in samplers/optimizers to be implemented in the same python module.
The old approach (method named ``evaluate''  is still available) but a new approach where a method named like the function is implemented.

@wangcj05 @mandd @idaholab/raven-team This is a modification I did in my local version and I thought it could be useful. Feel free to close/discard this feature/PR. 

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by alfoa** [comment]
@wangcj05 @mandd @joshua-cogliati-inl the windows machine does not seem to work

**Comment by wangcj05** [comment]
@alfoa FYI, our HPC is under maintenance right now. This outage will last until May 6. Please let me know if you need to merge this PR before that.

**Comment by alfoa** [comment]
> @alfoa FYI, our HPC is under maintenance right now. This outage will last until May 6. Please let me know if you need to merge this PR before that.

No problem for me. I can use my local branch for now. Let me know if you have comments on this in the meanwhile (that I can address offline).

Thanks a lot

File: ravenframework/Samplers/CustomSampler.py [hunk 1]
```diff
@@ -153,10 +154,16 @@ def _localGenerateAssembler(self,initDict):
...
       # check if the correct method is present
-      if "evaluate" not in self.funcDict[key].availableMethods():
-        self.raiseAnError(IOError, f'Function {self.funcDict[key].name} does not contain a method named "evaluate". It must be present if this needs to be used in a Sampler!')
+      if val not in initDict['Functions'][val].availableMethods():
+        if "evaluate" not in initDict['Functions'][val].availableMethods():
+          self.raiseAnError(IOError, f'Function {initDict["Functions"][val].name} does contain neither a method named "{val}" nor "evaluate". '
+                            'It must be present if this needs to be used in a Sampler!')
+      else:
+        mName = val
+      self.funcDict[key] = fPointer(mName, initDict['Functions'][val])
```
Comment (wangcj05) [inline ravenframework/Samplers/CustomSampler.py:166]: Can these lines be handled in the Sampler base class?

File: ravenframework/Samplers/EnsembleForward.py [hunk 2]
```diff
@@ -167,10 +168,16 @@ def _localGenerateAssembler(self, initDict):
...
       # check if the correct method is present
-      if "evaluate" not in self.funcDict[key].availableMethods():
-        self.raiseAnError(IOError, f'Function {self.funcDict[key].name} does not contain a method named "evaluate". It must be present if this needs to be used in a Sampler!')
+      if val not in availableFunc[val].availableMethods():
+        if "evaluate" not in availableFunc[val].availableMethods():
+          self.raiseAnError(IOError, f'Function {availableFunc[val].name} does contain neither a method named "{val}" nor "evaluate". '
+                            'It must be present if this needs to be used in a Sampler!')
+      else:
+        mName = val
+      self.funcDict[key] = fPointer(mName, availableFunc[val])
```
Comment (wangcj05) [inline ravenframework/Samplers/EnsembleForward.py:180]: Can these lines be handled in the Sampler base class?

**Review by wangcj05 — CHANGES_REQUESTED** [review]
@alfoa I have some comments for you to consider.

File: ravenframework/Samplers/CustomSampler.py (hunk 1 above)
Comment (alfoa) [reply ravenframework/Samplers/CustomSampler.py:166]: Unfortunately it seems that the CustomSampler and the EnsembleForward are "special" cases and they take care of they functions directly. I did not add this "exception" with this implementation but it was like that already unfortunately.

File: ravenframework/Samplers/EnsembleForward.py (hunk 2 above)
Comment (alfoa) [reply ravenframework/Samplers/EnsembleForward.py:180]: Unfortunately it seems that the CustomSampler and the EnsembleForward are "special" cases and they take care of they functions directly. I did not add this "exception" with this implementation but it was like that already unfortunately.

**Review by wangcj05 — APPROVED** [review]
changes are good.

**Comment by wangcj05** [comment]
Checklist is good, PR can be merged.

**Comment by wangcj05** [comment]
@alfoa Thanks for your contribution, I have merged your PR.

**Pull request merged**